### PR TITLE
Standardize inputs and refactor workflows

### PR DESCRIPTION
**Pull request by ChipWolf**
This PR refines several things:

- Various grammar/formatting nitpicks
- Removal of various redundant workflows, consolidated into fewer, easier to maintain workflows
- Replacing manual download commands with the `robinraju/release-downloader` action, so the action leverages the runner's `GITHUB_TOKEN` against the API as standard
- Grant execute permission to `gradlew` files and removed redundant workflow steps
- Added caching for CI builds and mc/modloader downloads

### CODE REVIEW

**Comment by 3arthqu4ke** [comment]
Looks awesome, do you think there is a way to provide caching for the release downloads?

**Comment by ChipWolf** [comment]
> Looks awesome, do you think there is a way to provide caching for the release downloads?

That's what I've done in our repo, I just need to get it into this branch

Only thing I need to consider is that many people (including ourselves) are using custom runners like Buildjet/Blacksmith/WarpBuild/Depot etc., and some of them have their own caching actions, so I should probably allow that to be provided as an input

Setup Java as you can see I moved out of the action because that seems to be the standard for most other workflows including some of GitHub's own, and it gives people some flexibility (especially if they're running this as part of a build & test job which would require Java already)

**Comment by ChipWolf** [comment]
Question @3arthqu4ke, why does `x11-xserver-utils` need to be installed?
I note our workflows appear to run fine without it, is there some edge cases that need it?

**Comment by ChipWolf** [comment]
Alright, that's the majority of what I wanted to achieve done

The scope has increased somewhat and I'll update my docs, but I've pushed regardless

File: .github/workflows/release-all.yml
Comment (ChipWolf) [inline .github/workflows/release-all.yml:1]: [DRY] redundant. Squashed into `flex-build` / `run-matrix`

File: .github/workflows/run-gametests-latest.yml
Comment (ChipWolf) [inline .github/workflows/run-gametests-latest.yml:1]: Moved to `run-gametests.yml`

File: .github/workflows/run-latest.yml
Comment (ChipWolf) [inline .github/workflows/run-latest.yml:1]: Not required

File: .github/workflows/run-local-test.yml
Comment (ChipWolf) [inline .github/workflows/run-local-test.yml:1]: TODO: feature parity with `run-specific-reusable`

File: .github/workflows/run-matrix-with-hmc.yml
Comment (ChipWolf) [inline .github/workflows/run-matrix-with-hmc.yml:1]: [DRY] redundant. Squashed into `run-matrix`

File: .github/workflows/run-matrix.yml
Comment (ChipWolf) [inline .github/workflows/run-matrix.yml:1]: TODO: add `workflow_dispatch.inputs` for build subsets per `build-all`

File: .github/workflows/run-matrix.yml [hunk 1]
```diff
@@ -1,98 +1,169 @@
...
+
+"on":
   push:
+    tags:
+      - '**'
+    branches:
+      - main
   pull_request:
   workflow_dispatch:
-  workflow_call:
```
Comment (ChipWolf) [inline .github/workflows/run-matrix.yml:6]: This workflow is not called by other workflows

File: .github/workflows/run-specific-reusable.yml
Comment (ChipWolf) [inline .github/workflows/run-specific-reusable.yml:1]: [DRY] redundant. Squashed into `run-local-test`

File: .github/workflows/run-specific-with-download-artifact-reusable.yml
Comment (ChipWolf) [inline .github/workflows/run-specific-with-download-artifact-reusable.yml:1]: [DRY] redundant. See `action.yml`.

`mc-runtime-test` can be set to none, the artifact can be downloaded to the cwd by the parent workflow, per `run-matrix`.

File: .github/workflows/run-specific-xvfb.yml
Comment (ChipWolf) [inline .github/workflows/run-specific-xvfb.yml:1]: [DRY] redundant. `xvfb` is an input for `action.yml`

File: .github/workflows/run-specific.yml
Comment (ChipWolf) [inline .github/workflows/run-specific.yml:1]: [DRY] redundant. See `action.yml`

File: .github/workflows/run-test-with-action.yml
Comment (ChipWolf) [inline .github/workflows/run-test-with-action.yml:1]: Moved to `run-local-test`

File: .github/workflows/verify-wrapper.yml
Comment (ChipWolf) [inline .github/workflows/verify-wrapper.yml:1]: Moved to `validation`

File: 1_12/gradlew
Comment (ChipWolf) [inline 1_12/gradlew:1]: `git update-index --chmod=+x gradlew`

File: api/build.gradle [hunk 2]
```diff
@@ -37,7 +37,7 @@ afterEvaluate {
                     name = 'GithubPagesMaven'
                     url = System.getenv('DEPLOY_TO_GITHUB_PACKAGES_URL')
                     credentials {
-                        username = System.getenv('GITHUB_USER')
+                        username = System.getenv('GITHUB_ACTOR')
```
Comment (ChipWolf) [inline api/build.gradle:40]: Per Gradle/GitHub docs

**Comment by 3arthqu4ke** [comment]
> I was a little overwhlemed by the huge number of actions in this repo, many of which don't appear to be used, so I'd like to consolidate those in this PR also

Yes many can just be thrown out.

> Question @3arthqu4ke, why does `x11-xserver-utils` need to be installed? I note our workflows appear to run fine without it, is there some edge cases that need it, or is it perhaps already installed in the runner?

This is an edge case for 1.12.2, with lwjgl-2.9.4 and xvfb, I am not sure if older lwjgl versions are affected.
In that version lwjgl crashes because it checks `xrandr` for displays, which is not installed on the runners, `x11-xserver-utils` provides it.
[https://github.com/3arthqu4ke/mc-runtime-test/actions/runs/11713260682/job/32674624091#step:21:3209](https://github.com/3arthqu4ke/mc-runtime-test/actions/runs/11713260682/job/32674624091#step:21:3209)

> Also, I note there's effectively 2 different "products" in this repo, the `mc-runtime-test` mod, and the action by the same name (which primary function is to headlessmc, but depends on `mc-runtime-test`), do you reckon there might be some value in the mod being in its own repo?

Hmmm, good question. Value in putting the mod in its own repository would probably be that it would make workflows in this repository much faster, because the run workflow does not need to build every single mod version every time.

On the other hand if I release a new version of the mod it means I also need to release a new version of the action, having it in one place makes that easier, but that is a nitpick.

**Comment by ChipWolf** [comment]
> Hmmm, good question. Value in putting the mod in its own repository would probably be that it would make workflows in this repository much faster, because the run workflow does not need to build every single mod version every time.
> 
> On the other hand if I release a new version of the mod it means I also need to release a new version of the action, having it in one place makes that easier, but that is a nitpick.

I could set up Renovate bot to deal with automated dependency updates & Release Please to deal with automated version and releasing?

**Comment by ChipWolf** [comment]
> This is an edge case for 1.12.2, with lwjgl-2.9.4 and xvfb, I am not sure if older lwjgl versions are affected.
> In that version lwjgl crashes because it checks xrandr for displays, which is not installed on the runners, x11-xserver-utils provides it.
> [3arthqu4ke/mc-runtime-test/actions/runs/11713260682/job/32674624091#step:21:3209](https://github.com/3arthqu4ke/mc-runtime-test/actions/runs/11713260682/job/32674624091#step:21:3209)

I can't replicate this by the way

**Comment by ChipWolf** [comment]
- [ ] https://github.com/3arthqu4ke/mc-runtime-test/pull/22#discussion_r1862995369

- [ ] https://github.com/3arthqu4ke/mc-runtime-test/pull/22#discussion_r1862993324

**Comment by 3arthqu4ke** [comment]
> > This is an edge case for 1.12.2, with lwjgl-2.9.4 and xvfb, I am not sure if older lwjgl versions are affected.
> > In that version lwjgl crashes because it checks xrandr for displays, which is not installed on the runners, x11-xserver-utils provides it.
> 
>     * ~Got it, makes sense. I’ll update the workflow to install `x11-xserver-utils` only when it’s needed, so it doesn’t add unnecessary overhead for other versions. We could also try this variable: https://github.com/LWJGL/lwjgl/blob/2df01dd762e20ca0871edb75daf670ccacc89b60/src/java/org/lwjgl/opengl/LinuxDisplay.java#L214~
> 
>     * I can't replicate this; see https://github.com/marcusk-studio/mc-runtime-test/actions/runs/12085945442/job/33704307459

I have tried the LWJGL_DISABLE_XRANDR variable and it did not work, lwjgl tries to fallback to XF86VIDMODE which also fails and then crashes.

Odd, I will look into this. Your runner is newer than the one used for my workflow but besides that I have no clue.

> 
> Honestly, I’m leaning toward splitting them. The mod feels pretty repetitive right now _(lots of duplicated code between version folders)_, and separating it would make it easier to DRY up without affecting the action workflows. Plus, I’m not a Java dev, so having it in its own space would make the action easier to work on.
> 

Yeah, I also tend to splitting them, I have plans for the duplicate code. While supporting many different minecraft and modloader versions is difficult, it could be possible with a Java Pre-processor https://github.com/3arthqu4ke/mc-runtime-test/issues/8

**Comment by 3arthqu4ke** [comment]
>   I can't replicate this; see https://github.com/marcusk-studio/mc-runtime-test/actions/runs/12085945442/job/33704307459

https://github.com/marcusk-studio/mc-runtime-test/actions/runs/12085945442/job/33704307750#step:6:583
Here you are not specifying -Dhmc.check.xvfb=true and you are in offline mode so headlessmc will automatically run with -lwjgl.

In the other workflow the -lwjgl flag is used.
https://github.com/marcusk-studio/mc-runtime-test/actions/runs/12085945442/job/33704307459#step:6:569

**Comment by 3arthqu4ke** [comment]
https://github.com/marcusk-studio/mc-runtime-test/blob/e42dbcc8274572ba4ec533a6068cc7b6eb5d3732/.github/workflows/run-matrix.yml#L178

I think this is the wrong way around? If you use xvfb you do not want to add -lwjgl?

**Comment by ChipWolf** [comment]
> https://github.com/marcusk-studio/mc-runtime-test/blob/e42dbcc8274572ba4ec533a6068cc7b6eb5d3732/.github/workflows/run-matrix.yml#L178
> 
> I think this is the wrong way around? If you use xvfb you do not want to add -lwjgl?

I inverted the logic on this test run, same thing https://github.com/marcusk-studio/mc-runtime-test/actions/runs/12104807693

**Comment by 3arthqu4ke** [comment]
> I inverted the logic on this test run, same thing https://github.com/marcusk-studio/mc-runtime-test/actions/runs/12104807693
> 

The lwjgl flag is still there in the xvfb-run.
https://github.com/marcusk-studio/mc-runtime-test/actions/runs/12104807693/job/33748650272#step:6:104
image

> In the name of progress, I'm inclined to polish and get this PR out of WIP so we can get it merged and make more progress. I don't want any more scope creep on this PR.
> 
> I'll add Release-Please and Renovate in a separate PR

:+1:

**Comment by ChipWolf** [comment]
The issue with the `lwjgl` flag in `xvfb-run` came down to how GitHub Actions handles its syntax. Originally, I tried this:

```yaml
${{ matrix.xvfb && '' || '-lwjgl' }}
```

Normally, you'd expect this to work like a ternary operator:

```yaml
${{ condition && 'if true' || 'if false' }}
```

But in this case, even when `matrix.xvfb` was `true`, the empty string (`''`) was treated as `false`. So the `|| '-lwjgl'` part kicked in. Turns out, GitHub Actions doesn’t support a proper ternary operator, and that empty string messed things up.

To fix it, I flipped the logic:

```yaml
${{ !matrix.xvfb && '-lwjgl' || '' }}
```

This works because:
1. `!matrix.xvfb` checks if `matrix.xvfb` is `false`.
2. If it is, you get `'-lwjgl'`.
3. If not, it falls back to `''`.

Now the empty string is in the right spot, and everything works as expected. It’s a simple tweak, but it gets around the weirdness with GitHub Actions’ syntax.

**Comment by 3arthqu4ke** [comment]
On timeout: headlessmc handles retries and retries on timeout would also be great. Just opening the discussion to the question if handling timeout in headlessmc would make sense, or handling retries in mc-runtime-test would make sense.

The reason I put retries on the headlessmc side for now was to distinguish between hmc configuration errors and minecraft errors but not sure.

**Comment by 3arthqu4ke** [comment]
Ah and this time the 1.19.3 test just timed out, just as I wrote about the 1.19.3 flakyness.
I think I will need to look into this on the mc-runtime-test mod side as well.

**Comment by 3arthqu4ke** [comment]
> This is just an observation. I dont think this is necessarily bad because I wouldnt know how to do it differently while having a cleanup step. I like that the artifacts are deleted, but if a test fails we need to re-run all tests, instead of the cancelled/failed ones.

Actually the only idea I just had would be to have a conditional confirmation step before the cleanup step if the run step failed.
But only if you want to, this is just a small optimization and not really in the scope.

**Comment by ChipWolf** [comment]
> This is just an observation. I dont think this is necessarily bad because I wouldnt know how to do it differently while having a cleanup step. I like that the artifacts are deleted, but if a test fails we need to re-run all tests, instead of the cancelled/failed ones.

Perhaps we should only clean up artifacts from passed tests

I'm working on selective builds right now based on the content of a PR, so in those instances I believe we shouldn't remove the artifacts either

**Comment by ChipWolf** [comment]
Alright this is about as polished as I can manage for now @3arthqu4ke, I'll raise another PR after this is in for the dependency updates and semantic versioning etc.

File: action.yml [hunk 3]
```diff
@@ -1,139 +1,133 @@
...
+        repository: 3arthqu4ke/mc-runtime-test
+        tag: "2.4.2"
+        fileName: mc-runtime-test-${{ inputs.mc }}-2.4.2-${{ inputs.mc-runtime-test }}-release.jar
+        out-file-path: run/mods
+
+    - if: inputs.fabric-api != 'none'
+      name: Download fabric-api v${{ inputs.fabric-api }}
+      run: >
+        wget -O run/mods/fabric-api-${{ inputs.fabric-api }}+${{ inputs.mc }}.jar
+        https://maven.fabricmc.net/net/fabricmc/fabric-api/fabric-api/${{ inputs.fabric-api }}+${{ inputs.mc }}/fabric-api-${{ inputs.fabric-api }}+${{ inputs.mc }}.jar
```
Comment (ChipWolf) [inline action.yml:105]: DRY this, we should avoid referencing the filename twice.

File: action.yml [hunk 4]
```diff
@@ -1,139 +1,133 @@
...
       shell: bash
-    - name: LS mods dir
-      run: ls run/mods
+
+    - if: inputs.fabric-gametest-api != 'none'
+      name: Download fabric-gametest-api v${{ inputs.fabric-gametest-api }}
+      run: >
+        wget -O run/mods/fabric-gametest-api-v1-${{ inputs.fabric-gametest-api }}.jar
+        https://maven.fabricmc.net/net/fabricmc/fabric-api/fabric-gametest-api-v1/${{ inputs.fabric-gametest-api }}/fabric-gametest-api-v1-${{ inputs.fabric-gametest-api }}.jar
```
Comment (ChipWolf) [inline action.yml:112]: DRY this, we should avoid referencing the filename twice.

**Comment by 3arthqu4ke** [comment]
Could you rebase on top of main real quick, please? I think I fixed the 1.19 issues, that way I can run the workflows for this PR without having to restart them if 1.19 fails.

**Review by 3arthqu4ke — APPROVED** [review]
Thanks a lot, this is awesome!

**Review by 3arthqu4ke — APPROVED** [review]

**Comment by 3arthqu4ke** [comment]
Workflows now no longer fail fast if something in the run step goes wrong, but I guess that was to be expected?

**Comment by ChipWolf** [comment]
> Workflows now no longer fail fast if something in the run step goes wrong, but I guess that was to be expected?

Yeah that was a last minute choice
My thinking is if you've changed various things you probably want to see all the problems in one go

**Comment by ChipWolf** [comment]
> Could you rebase on top of main real quick, please? I think I fixed the 1.19 issues, that way I can run the workflows for this PR without having to restart them if 1.19 fails.

Will do

**Comment by 3arthqu4ke** [comment]
> > Could you rebase on top of main real quick, please? I think I fixed the 1.19 issues, that way I can run the workflows for this PR without having to restart them if 1.19 fails.
> 
> Will do

No need to, workflows ran successfully, will merge now, Just wanted clarification on my comment. Perfekt, great Work!

**Comment by 3arthqu4ke** [comment]
> Alright this is about as polished as I can manage for now @3arthqu4ke, I'll raise another PR after this is in for the dependency updates and semantic versioning etc.
> 
> I'm sure there'll be some teething issues, but it'll be simple enough to squash them now

Should I just bump everything to 3.0.0 and release, or what did you have in mind?

**Comment by ChipWolf** [comment]
> > Alright this is about as polished as I can manage for now @3arthqu4ke, I'll raise another PR after this is in for the dependency updates and semantic versioning etc.
> > I'm sure there'll be some teething issues, but it'll be simple enough to squash them now
> 
> Should I just bump everything to 3.0.0 and release, or what did you have in mind?

Let me get the renovate/release-please PR in then we can run the 3.0.0 release automatically
